### ISSUE TITLE
[no-unreachable] switch statement

### ISSUE THREAD

**Issue by galambalazs** [eslint/eslint#5601]
When `break` is used consistently to avoid programmer errors in the future, eslint will warn about having `return` & `break` (the `break` being unreachable).

I think this is similar to the (redundant) trailing commas where they are *preventing future errors* so this should be allowed:

```js
switch (cond) {
  case 'one':
    if (cond1) 
      return 'denied';
    else 
      return 'good'
  break; // <--- this will trigger the warning of unreachable code
  case 'two':
    simple code..
    just statements..
  break;
  default: 
    return 'unexpected';
}
```

(note: it may also be a stylistic/formatting preference that all cases have break pairs)

**Comment by galambalazs** [comment]
note: JSLint/JSHint has a special case for this too: `"Unreachable 'break' after 'return'."`
(I think it's on warning by default on JSHint)

**Comment by platinumazure** [comment]
I can see why it would be useful, but it feels like an awfully specific case and you can use `no-fallthrough` to catch missing `break` statements. I'm :-1: on this proposal.

**Comment by michaelficarra** [comment]
Agreed with @platinumazure. Just make sure you have `no-fallthrough` in the same config and you won't need this insurance.

**Comment by ilyavolodin** [comment]
I agree. If you want consistency either disable this rule around switch cases or just turn it off globally. `break` in your example is unreachable, so the rule is doing the right thing and warning you about it.
Closing since this is working as intended.

**Comment by galambalazs** [comment]
This may work as you guys intended, but then your intentions simply don't cover all the cases.
As with other rules it not always an on/off binary option, sometimes you have more control.

Disabling globally or even around switch is **not a solution**.

Consider the following example:

```javascript
switch (cond) {
  case 'one':
    if (cond1) 
      return 'denied';
    else 
      return 'good'
  break; // <--- doesn't make sense to force you to remove this line
  case 'two':
    if (cond1) 
      return 'denied';
    else 
      return 'good'
    leftAfterRefactor() // <---- this is a valid place where it should warn you
  break;
}
```

In the first case it makes no sense for ESLint to enforce the rule, but if you turn it off you lose the helpful warnings e.g. in the second case.

**Comment by michaelficarra** [comment]
@galambalazs Removing the `break` was the suggested solution.

**Comment by galambalazs** [comment]
Should you also remove all trailing commas? They are dead code too, you know. 
Oh wait, but there's an option for that. Interesting... :)

So eslint *can* be about having sensible options.

**Comment by platinumazure** [comment]
I'm not sure I get the problem. You turn on `no-unreachable` and it correctly flags an unreachable break statement. You state, as your reason for wanting to allow those statements not to be flagged as unreachable, that you don't want there to be accidental fallthrough if someone removes the return statements preceding the break statement. We recommend using `no-fallthrough` which is designed for exactly that situation. Why won't that work for you?

**Comment by ilyavolodin** [comment]
> This may work as you guys intended, but then your intentions simply don't cover all the cases. :)

We never intended to cover all the cases. That's simply not possible. You use-case seems to be an edge case. I understand that's your preferred style, but I haven't seen this pattern used widely and I'm not sure I understand why you would want to do that. If you can point us to a popular style-guide or very popular library using this patter, we would be happy to revisit this discussion. If not - you always have an option of creating a plugin that would flag/not flag exactly the cases that you prefer.

**Comment by HarryPehkonen** [comment]
This is the solution that I use, and my story of the whole issue.  I'm using eslint v.2.9.0.

For what it's worth, if I don't see "break" at the end of every switch/case, I get concerned.  From my C-days, I am used to adding /* NOTREACHABLE \*/ or /* FALLTHROUGH */ inside switch/case if I'm intending something specific.  Having support for something like this ("case" by "case") seems beneficial.

Specifying:
```
switch (x) {
case "something":
    /* eslint no-unreachable: 0 */
    return x;
    break;
    /* eslint no-unreachable: 2 */
```

... doesn't help because it's the final rule that applies to the whole .js file (the first no-unreachable is, in effect, ignored).  This is, of course, by design.

Adding either eslint-disable-line or eslint-disable-next-line inside /* */ comments:
```
return;
/* eslint-disable-line no-unreachable */
/* eslint-disable-line no-unreachable */ break; /* eslint-disable-line no-unreachable */
/* eslint-disable-line no-unreachable */
```
...doesn't help either (yes, I tried all four possible places for it).

These, however, work (the use of // instead of /* */ for the comments is required -- is that a bug?):
```
return;
/* eslint-disable no-unreachable */
break;
/* eslint-enable no-unreachable */
```
```
return;
// eslint-disable-next-line no-unreachable
break;
```

```
return;
break; // eslint-disable-line no-unreachable
```

I hope that helps.

Harry.
